internal/orchestrator/reorg_handler.go

```diff
@@ -0,0 +1,187 @@
+package orchestrator
+
+import (
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	config "github.com/thirdweb-dev/indexer/configs"
+	"github.com/thirdweb-dev/indexer/internal/common"
+	"github.com/thirdweb-dev/indexer/internal/metrics"
+	"github.com/thirdweb-dev/indexer/internal/rpc"
+	"github.com/thirdweb-dev/indexer/internal/storage"
+	"github.com/thirdweb-dev/indexer/internal/worker"
+)
+
+type ReorgHandler struct {
+	rpc              rpc.Client
+	storage          storage.IStorage
+	triggerInterval  int
+	blocksPerScan    int
+	lastCheckedBlock *big.Int
+	worker           *worker.Worker
+}
+
+const DEFAULT_REORG_HANDLER_INTERVAL = 1000
+const DEFAULT_REORG_HANDLER_BLOCKS_PER_SCAN = 100
+
+func NewReorgHandler(rpc rpc.Client, storage storage.IStorage) *ReorgHandler {
+	triggerInterval := config.Cfg.ReorgHandler.Interval
+	if triggerInterval == 0 {
+		triggerInterval = DEFAULT_REORG_HANDLER_INTERVAL
+	}
+	blocksPerScan := config.Cfg.ReorgHandler.BlocksPerScan
+	if blocksPerScan == 0 {
+		blocksPerScan = DEFAULT_REORG_HANDLER_BLOCKS_PER_SCAN
+	}
+	return &ReorgHandler{
+		rpc:              rpc,
+		storage:          storage,
+		worker:           worker.NewWorker(rpc),
+		triggerInterval:  triggerInterval,
+		blocksPerScan:    blocksPerScan,
+		lastCheckedBlock: getInitialCheckedBlockNumber(storage, rpc.ChainID),
+	}
+}
+
+func getInitialCheckedBlockNumber(storage storage.IStorage, chainId *big.Int) *big.Int {
+	bn := big.NewInt(int64(config.Cfg.ReorgHandler.FromBlock))
+	if !config.Cfg.ReorgHandler.ForceFromBlock {
+		storedFromBlock, err := storage.OrchestratorStorage.GetLastReorgCheckedBlockNumber(chainId)
+		if err != nil {
+			log.Debug().Err(err).Msgf("Error getting last reorg checked block number, using configured: %s", bn)
+			return bn
+		}
+		if storedFromBlock.Sign() <= 0 {
+			log.Debug().Msgf("Last reorg checked block number not found, using configured: %s", bn)
+			return bn
+		}
+		log.Debug().Msgf("Last reorg checked block number found, using: %s", storedFromBlock)
+		return storedFromBlock
+	}
+	log.Debug().Msgf("Force from block reorg check flag set, using configured: %s", bn)
+	return bn
+}
+
+func (rh *ReorgHandler) Start() {
+	interval := time.Duration(rh.triggerInterval) * time.Millisecond
+	ticker := time.NewTicker(interval)
+
+	log.Debug().Msgf("Reorg handler running")
+	go func() {
+		for range ticker.C {
+			lookbackFrom := new(big.Int).Add(rh.lastCheckedBlock, big.NewInt(int64(rh.blocksPerScan)))
```

Comment:
wouldn't this skip one batch at the beginning?

Reply:
No, it's a lookback, so it gets blocks lower than `lookbackFrom `